### PR TITLE
 Solve skeletal inverse kinematic poses with constraints

### LINKED ISSUE #6039: Solve poses for full body VR tracking with Many Bone IK
### Describe the project you are working on

A open source social vr game made with Godot Engine.

We wish to provide full body tracking with 3 body sensors and the ground or up and more than 10 body sensors.

### Describe the problem or limitation you are having in your project

Godot Engine 4's multiple ik systems are broken.

https://twitter.com/wojtekpil/status/1609667214998650883?s=20

Provide a general purpose full body inverse kinematic system that is suitable for full body tracking and everything else inverse kinematics related.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Many Bone IK is a custom inverse kinematics system solver for multi-chain skeletons and with constraints.

## Godot Engine Many Bone IK authors

GitHub usernames are indicated in parentheses.

- K. S. Ernest (iFire) Lee (fire)
- Eron Gjoni (EGjoni)
- rafallus Rafael M. G. (rafallus)
- lyuma (lyuma)

## Initial Project

This project is originally a port from a Java project by Eron

    Eron Gjoni (EGjoni) https://github.com/EGjoni/Everything-Will-Be-IK

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

DRAFT PROPOSAL as of 2022-01-03.

image

image

# Abstract

2. provided IK COPY targets per bone via "QCP" solver
4. provided KUSUDAMA IK LIMIIT forward axial rotation per bone with a 0-1 current slider
5. provided KUSUDAMA IK LIMIT spherical (360) rotation per bone via 0 to 30 limit cones
1. provided root motion solving
6. included stabilization passes
7. provided prioritization of rotation axes
6. provided joint weights
8. Per bone dampening
9. provided spherical KUSUDAMA IK LIMIT shader visualization
10. provided axial KUSUDAMA IK LIMIT visualization
11. Avoids the modification stack problem mentioned in earlier pull requests in July. ManyBoneIK is a node3d.
3. Allows arbitrary number of iterations per frame
4. Provides a constraint mode than allows editing of constraints without the solve struggling
5. Provides a filterable inspector with a bone tree.

## Unfinished work

1. provide a better ui for defining 360 spherical rotation like drilling from a sphere center point (vector3) and then adding a radius (euler angles) 
3. provide a pallete of presets for common humanoid joints
2. provide draggable handles for spherical KUSUDAMA LIMIT similar to the spotlight3d set to a distance of 1 scaled by the bone length
6. draggable handles for the axial limits
7. the ui is too complicated needs ui work
8. film an demonstration video
9. Make ManyBoneIK a node? (trouble with gizmos.)
10. Finish the demo https://github.com/V-Sekai-fire/TEST_many_bone_ik

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This code is not trivial. The many bone ik solver took several years since 2016 to get the ik solver to this point.

### Is there a reason why this should be core and not an add-on in the asset library?

Reduz asked me to write a proposal because I had working code for this in a c++ module at https://github.com/V-Sekai/many_bone_ik

### PR DESCRIPTION
**Pull request by fire**
Main work will be in https://github.com/V-Sekai/many_bone_ik

**Proposal for Godot Engine 4.x** 

Fixes: https://github.com/godotengine/godot-proposals/issues/6039

Adds another option to https://github.com/godotengine/godot/issues/69752

Supersedes https://github.com/godotengine/godot/pull/63588

Part of V-Sekai https://github.com/v-sekai/v-sekai-game

# A list of todos

- [ ] Fully constraint test project from https://github.com/godotengine/godot/issues/69752
- [ ] Add cone handle adjust radius and spherical surface position with a distance of 1 from the spotlight3d
- [ ] Do a dance.
- [ ] BETTER UIUX BLOCKER
- [x] Mark experimental
- [x] ~~Delay painfulness parameter, but call it something else.~~
- [x] Fix increasing the cone count resetting the inspector

### CODE REVIEW

**Comment by fire** [comment]
Wanted to sense the mood on the enhancement and if it's positive, try moving the custom code parts into the core.

File: modules/many_bone_ik/doc_classes/IKBone3D.xml [hunk 1]
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="IKBone3D" inherits="Resource" version="4.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../doc/class.xsd">
+	<brief_description>
+	</brief_description>
+	<description>
+	</description>
+	<tutorials>
+	</tutorials>
+	<methods>
+		<method name="get_constraint" qualifiers="const">
```
Comment (fire) [inline modules/many_bone_ik/doc_classes/IKBone3D.xml:10]: Isn't this the godot aligned bone transform? Rename?

File: modules/many_bone_ik/doc_classes/IKRay3D.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="IKRay3D" inherits="RefCounted" version="4.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../doc/class.xsd">
+	<brief_description>
+	</brief_description>
+	<description>
+	</description>
+	<tutorials>
+	</tutorials>
+	<methods>
+		<method name="heading">
```
Comment (fire) [inline modules/many_bone_ik/doc_classes/IKRay3D.xml:10]:
```suggestion
		<method name="get_heading">
```

File: modules/many_bone_ik/doc_classes/IKRay3D.xml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+			<return type="Vector3" />
+			<description>
+			</description>
+		</method>
+		<method name="intersects_plane">
```
Comment (fire) [inline modules/many_bone_ik/doc_classes/IKRay3D.xml:15]:
```suggestion
		<method name="get_intersection_plane">
```

File: modules/many_bone_ik/doc_classes/IKRay3D.xml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+			<return type="Vector3" />
+			<param index="0" name="a" type="Vector3" />
+			<param index="1" name="b" type="Vector3" />
+			<param index="2" name="c" type="Vector3" />
+			<description>
+			</description>
+		</method>
+		<method name="scaled_projection">
```
Comment (fire) [inline modules/many_bone_ik/doc_classes/IKRay3D.xml:23]:
```suggestion
		<method name="get_scaled_projection">
```

File: modules/many_bone_ik/src/ik_bone_segment.h [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	PackedVector3Array target_headings;
+	PackedVector3Array tip_headings;
+	PackedVector3Array tip_headings_uniform;
+	Vector<real_t> heading_weights;
+	BoneId many_bone_ik_tip_bone = -1;
+	int32_t idx_eff_i = -1, idx_eff_f = -1;
+	Skeleton3D *skeleton = nullptr;
+	bool pinned_descendants = false;
+	real_t previous_deviation = 0;
+	int32_t default_stabilizing_pass_count = 1; // Move to the stabilizing pass to the ik solver.
```
Comment (fire) [inline modules/many_bone_ik/src/ik_bone_segment.h:66]: Expose this once for the entire instance of the ik solver

File: modules/many_bone_ik/src/ik_effector_3d.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+	Vector3 bone_origin_relative_to_skeleton_origin = p_for_bone->get_bone_direction_global_pose().origin;
+	int32_t index = p_index;
+	p_headings->write[index] = tip_xform_relative_to_skeleton_origin.origin - bone_origin_relative_to_skeleton_origin;
+	index++;
+	double scale_by = 1.0f;
+	if (p_is_uniform) {
+		double distance = target_relative_to_skeleton_origin.origin.distance_to(bone_origin_relative_to_skeleton_origin);
+		scale_by = MAX(1.0, distance);
+	}
+	// Scale by means when two things that are very far away, they may close in physical space but they get a higher weighting.
```
Comment (fire) [inline modules/many_bone_ik/src/ik_effector_3d.cpp:152]:
```suggestion
	// Scale by means when two things that are very far away, they may be close in physical space but they get a higher weighting.
```

File: modules/many_bone_ik/src/ik_kusudama.cpp [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+bool IKKusudama::is_axially_constrained() {
+	return axially_constrained;
+}
+
+bool IKKusudama::is_orientationally_constrained() {
+	return orientationally_constrained;
+}
+
+void IKKusudama::disable_orientational_limits() {
+	this->orientationally_constrained = false;
```
Comment (fire) [inline modules/many_bone_ik/src/ik_kusudama.cpp:194]: remove this-> from java port.

File: modules/many_bone_ik/src/ik_kusudama.h [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+	 * Defined as some Angle in radians about the limiting_axes Y axis, 0 being equivalent to the
+	 * min_axial_angle
+	 */
+	real_t range_angle = Math_TAU;
+
+	bool orientationally_constrained = false;
+	bool axially_constrained = false;
+
+	/**
+	 * Get the swing rotation and twist rotation for the specified axis. The twist rotation represents the rotation around the specified axis. The swing rotation represents the rotation of the specified
```
Comment (fire) [inline modules/many_bone_ik/src/ik_kusudama.h:86]: Move into xml documentation.

File: modules/many_bone_ik/src/ik_bone_3d.h [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	// The origin is the origin of the bone direction transform
+	// Can be independent and should be calculated
+	// to keep -y to be the opposite of its bone forward orientation
+	// To avoid singularity that is ambiguous.
+	Ref<IKNode3D> constraint_orientation_transform = Ref<IKNode3D>(memnew(IKNode3D()));
+	Ref<IKNode3D> constraint_twist_transform = Ref<IKNode3D>(memnew(IKNode3D()));
+	Ref<IKNode3D> godot_skeleton_aligned_transform = Ref<IKNode3D>(memnew(IKNode3D())); // The bone's actual transform.
+	Ref<IKNode3D> bone_direction_transform = Ref<IKNode3D>(memnew(IKNode3D())); // Physical direction of the bone. Calculate Y is the bone up.
+
+	/*Ref<IKNode3D> __default_constraint_transform = Ref<IKNode3D>(memnew(IKNode3D()));
```
Comment (fire) [inline modules/many_bone_ik/src/ik_bone_3d.h:71]: TODO:

File: modules/many_bone_ik/doc_classes/ManyBoneIK3D.xml [hunk 10]
```diff
@@ -0,0 +1,245 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="ManyBoneIK3D" inherits="Node3D" version="4.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../../../doc/class.xsd">
+	<brief_description>
+		A full body inverse kinematics system.
+	</brief_description>
+	<description>
+		A full body inverse kinematics system.
```
Comment (fire) [inline modules/many_bone_ik/doc_classes/ManyBoneIK3D.xml:7]: Duplicate sentence.
```suggestion
```

**Comment by TheoJao** [comment]
> > How is this supposed to work related to #63588? Will eventually both get merged at different times or will one supercede the other?
> 
> Seems to me like this implementation is completely superior to the skeleton modification one, so it would probably supercede the skeleton modification system entirely.
> 
> I'm a bit sad that IK 3D wasn't given higher priority for 4.0's (and 4.1's) release, you can't really make a proper modern 3D game without inverse kinematics, it serves many critical roles. Yet 4.0 released with nothing but a completely broken SkeletonIK3D node implementation (thing cannot be used at all, it is broken beyond any sort of usability in 4.0.1, if they removed skeleton modification because it was broken, they should do the same for skeletonik3d, cuz it's broken too in just about every way. Hell, u even get an error just from adding the node, an actual error!)
> 
> I really wish this pull request would go through soon. I'm very new to the engine, and one of the contributing factors for me choosing godot was because I saw some very good IK implementations in videos about godot, but then I install the engine, start learning it and find out that IK is actually completely broken? It was very disappointing for me, it's critical for my project.

If you really need it, you can include this module into your project, however it's kind of hard to use without any documentation. You could also try an IK add on named Twisted IK 2 which you can find on itch.io. It's paid but you can download a demo to try it out before buying it.
I do hope an IK solution is added to Godot soon, it's been almost 2 years since IK got broken.

**Comment by fire** [comment]
I accidentally deleted the branch as we're reworking in https://github.com/V-Sekai/many_bone_ik.

We have this idea of moving inverse kinematics to the animation tree. Still working on a proposal.

https://gist.github.com/fire/1953bd6bb0e9cc3b964796317d23c702 has a draft for splitting the work into defining a standard approach for inverse kinematics in Godot Engine and implementing the many bone ik animation tree functionality.

**Comment by fire** [comment]
https://github.com/godotengine/godot/pull/77469 has a fix for the broken IK node in godot engine 4.0

**Pull request closed without merging**